File: ios-native/WordWheel/Views/BottomButtons.swift
import SwiftUI

/// Bottom action bar.
///
/// Drag-on-the-wheel commits the selected word automatically when the
/// gesture ends (see `LetterWheel`). That makes Submit and Backspace
/// buttons redundant — only the Hint button stays.
struct BottomButtons: View {
    let hintsLeft: Int
    let wordsTowardHint: Int
    let onHint: () -> Void

    var body: some View {
        HStack {
            HintButton(hintsLeft: hintsLeft,
                       wordsTowardHint: wordsTowardHint,
                       onHint: onHint)
        }
        .frame(maxWidth: .infinity)
        .padding(.vertical, 2)
    }
}

private struct HintButton: View {
    let hintsLeft: Int
    let wordsTowardHint: Int
    let onHint: () -> Void

    var body: some View {
        VStack(spacing: 3) {
            Button(action: onHint) {
                // Outer 78pt frame keeps the count badge sitting tight
                // against the 64pt gold disc — the previous 84pt left
                // a 10pt gap that read as "detached badge".
                ZStack(alignment: .topTrailing) {
                    ZStack {
                        Circle()
                            .fill(hintsLeft > 0
                                  ? Color(hex: 0xFFFFB400)
                                  : Color(hex: 0xB4282828))
                            .frame(width: 64, height: 64)
                            .shadow(color: .black.opacity(0.4), radius: 4, x: 0, y: 3)
                        Text("\u{1F4A1}")
                            .font(.system(size: 28))
                    }
                    .frame(width: 78, height: 78, alignment: .bottomLeading)

                    ZStack {
                        Circle()
                            .fill(hintsLeft > 0
                                  ? Color(hex: 0xFF32B450)
                                  : Color(hex: 0xFF787878))
                            .frame(width: 28, height: 28)
                            .overlay(Circle().stroke(Color.white, lineWidth: 2.5))
                            .shadow(color: .black.opacity(0.4), radius: 3, x: 0, y: 2)
                        Text("\(hintsLeft)")
                            .font(.system(size: 14, weight: .heavy))
                            .foregroundColor(.white)
                    }
                }
                .frame(width: 78, height: 78)
            }
            .buttonStyle(.plain)

            // Progress toward the next free hint (every 10 words found).
            ZStack(alignment: .leading) {
                RoundedRectangle(cornerRadius: 3)
                    .fill(Color(hex: 0xFF28283C))
                    .frame(width: 72, height: 5)
                let progress = CGFloat(min(wordsTowardHint, 10)) / 10.0
                if progress > 0 {
                    RoundedRectangle(cornerRadius: 3)
                        .fill(Color(hex: 0xFF50DC78))
                        .frame(width: 72 * progress, height: 5)
                }
            }
        }
    }
}

File: android-native/app/src/main/kotlin/com/wheelword/game/ui/BottomButtons.kt
package com.wheelword.game.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.ui.draw.shadow
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

/**
 * Bottom action bar.
 *
 * Drag-on-the-wheel is the primary submit gesture — releasing the drag
 * commits the selected word automatically (see [LetterWheel]). That
 * makes both Submit and Backspace buttons redundant. Only the Hint
 * button lives here now, centred and large.
 */
@Composable
fun BottomButtons(
    hintsLeft: Int,
    wordsTowardHint: Int,
    onHint: () -> Unit,
) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .padding(vertical = 2.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.Center,
    ) {
        HintButton(hintsLeft, wordsTowardHint, onHint)
    }
}

@Composable
private fun HintButton(hintsLeft: Int, wordsTowardHint: Int, onHint: () -> Unit) {
    Column(horizontalAlignment = Alignment.CenterHorizontally) {
        // Outer container snug against the 64dp gold disc — 78dp leaves
        // just enough room for the 28dp count badge at the top-right
        // corner with the badge sitting tight against the disc edge.
        // 84dp (the previous value) had ~10dp of awkward gap that read
        // as detached from the button.
        Box(
            modifier = Modifier.size(78.dp),
        ) {
            Box(
                modifier = Modifier
                    .size(64.dp)
                    .align(Alignment.BottomStart)
                    .shadow(elevation = 6.dp, shape = CircleShape)
                    .clip(CircleShape)
                    .background(
                        if (hintsLeft > 0) Color(0xFFFFB400)
                        else Color(0xB4282828)
                    )
                    .clickable(onClick = onHint),
                contentAlignment = Alignment.Center,
            ) {
                Text(
                    text = "\uD83D\uDCA1",
                    fontSize = 28.sp,
                )
            }
            // Count badge — anchored top-right of the outer 84dp box.
            // The 2.5dp white ring + drop shadow makes it pop against
            // the gold disc behind it so the digit is always readable.
            Box(
                modifier = Modifier
                    .align(Alignment.TopEnd)
                    .shadow(elevation = 4.dp, shape = CircleShape)
                    .size(28.dp)
                    .clip(CircleShape)
                    .background(
                        if (hintsLeft > 0) Color(0xFF32B450)
                        else Color(0xFF787878)
                    )
                    .border(width = 2.5.dp, color = Color.White, shape = CircleShape),
                contentAlignment = Alignment.Center,
            ) {
                Text(
                    text = hintsLeft.toString(),
                    color = Color.White,
                    fontSize = 14.sp,
                    fontWeight = FontWeight.Black,
                )
            }
        }
        // Progress toward the next free hint (every 10 words found).
        Box(
            modifier = Modifier
                .padding(top = 3.dp)
                .width(72.dp)
                .height(5.dp)
                .clip(RoundedCornerShape(3.dp))
                .background(Color(0xFF28283C)),
        ) {
            val progress = (wordsTowardHint.coerceAtMost(10)) / 10f
            if (progress > 0f) {
                Box(
                    modifier = Modifier
                        .fillMaxHeight()
                        .fillMaxWidth(progress)
                        .background(Color(0xFF50DC78)),
                )
            }
        }
    }
}
